pkg/pgxconn/implement.go

```diff
@@ -0,0 +1,102 @@
+package pgxconn
+
+import (
+	"time"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgproto3/v2"
+	"github.com/jackc/pgx/v4"
+)
+
+// rowWithMetric wraps the row returned by QueryRow() for metrics telemetry.
+type rowWithMetric struct {
+	row pgx.Row
+}
+
+func (w rowWithMetric) Scan(dest ...interface{}) error {
+	err := w.row.Scan(dest...)
+	if err != nil && err != pgx.ErrNoRows {
+		errorsTotal.With(promMethodLabel("query_row")).Inc()
+	}
+	return err
+}
+
+// rowsWithDuration wraps the Query() function with duration metric taken to complete the entire execution.
+type rowsWithDuration struct {
+	rows  pgx.Rows
+	start time.Time
+}
+
+func newRowsWithDuration(rows pgx.Rows, start time.Time) rowsWithDuration {
+	return rowsWithDuration{rows, start}
+}
+
+func (r rowsWithDuration) Next() bool {
+	return r.rows.Next()
+}
+
+func (r rowsWithDuration) Scan(dest ...interface{}) error {
+	return r.rows.Scan(dest...)
+}
+
+func (r rowsWithDuration) Err() error {
+	return r.rows.Err()
+}
+
+func (r rowsWithDuration) Close() {
+	defer func() { duration.With(promMethodLabel("query")).Observe(time.Since(r.start).Seconds()) }()
+	r.rows.Close()
+}
+
+func (r rowsWithDuration) CommandTag() pgconn.CommandTag {
+	return r.rows.CommandTag()
+}
+
+func (r rowsWithDuration) FieldDescriptions() []pgproto3.FieldDescription {
+	return r.rows.FieldDescriptions()
+}
+
+func (r rowsWithDuration) Values() ([]interface{}, error) {
+	return r.rows.Values()
+}
+
+func (r rowsWithDuration) RawValues() [][]byte {
+	return r.rows.RawValues()
+}
+
+// batchResultsWithDuration wraps the SendBatch and records the duration of the batch till it closed.
+type batchResultsWithDuration struct {
+	batch pgx.BatchResults
+	start time.Time
+}
+
+func newBatchResultsWithDuration(b pgx.BatchResults) batchResultsWithDuration {
+	return batchResultsWithDuration{
+		batch: b,
+		start: time.Now(),
+	}
+}
+
+func (w batchResultsWithDuration) Close() error {
+	defer func() {
+		duration.With(promMethodLabel("send_batch")).Observe(time.Since(w.start).Seconds())
+	}()
+	return w.batch.Close()
+}
+
+func (w batchResultsWithDuration) Exec() (pgconn.CommandTag, error) {
+	return w.batch.Exec()
+}
+
+func (w batchResultsWithDuration) Query() (pgx.Rows, error) {
+	rows, err := w.batch.Query()
+	return newRowsWithDuration(rows, time.Now()), err
```

Comment:
This doesn't seem in line with semantics to track the time from query start to rows close because `Query` for batches just moves to next result. My recommendation is to just skip this for now.

Reply:
This would also conflate two things: - we are tracking the `send_batch` time but `newRowsWithDuration` will track other time with `query` label which is semantically different so imho the best is to avoid tracking it.

Reply:
We can just track the whole batch from start to batch close

Reply:
However we might want to track for errors when scanning rows.